### PR TITLE
ci: add tests on s390x and ppc64le

### PR DESCRIPTION
**Pull request by Mohammed90**
Prepping for #3325

### CODE REVIEW

**Comment by francislavoie** [comment]
We should probably update the artifact names to match eachother. The new one is `linux_{arch}` but the old ones are `Linux_{shorthash}`.

**Comment by francislavoie** [comment]
@grooverdan might you have any insight as to why the IBM Z job is failing?

```
go: github.com/Masterminds/sprig/v3@v3.0.2: Get "https://proxy.golang.org/github.com/%21masterminds/sprig/v3/@v/v3.0.2.mod": tls: invalid signature by the server certificate: ECDSA verification failure
go: github.com/Masterminds/sprig/v3@v3.0.2: Get "https://proxy.golang.org/github.com/%21masterminds/sprig/v3/@v/v3.0.2.mod": tls: invalid signature by the server certificate: ECDSA verification failure
go: github.com/Masterminds/sprig/v3@v3.0.2: Get "https://proxy.golang.org/github.com/%21masterminds/sprig/v3/@v/v3.0.2.mod": tls: invalid signature by the server certificate: ECDSA verification failure
```

**Comment by Mohammed90** [comment]
To be clear, I'd like to set up testing infrastructure for those particular architectures to catch any failures that could impact the users on those architectures.

**Comment by grooverdan** [comment]
Ok. going to have a look today.

https://developer.ibm.com/linuxone/ and https://developer.ibm.com/linuxonpower/cloud-resources/ can be used to signup for Z/Power VMs for testing infrastructure.

I'm even doing this for Z to get my access to test this failure, at least see if this is a QEMU/hardware difference somewhere else in the software stack.

**Comment by grooverdan** [comment]
### Caddy passes tests on real s390x hardware

Rhel7.7, latest go-1.14.2

One quirk due to
<pre>
exec: "s390x-linux-gnu-gcc": executable file not found in $PATH
</pre>

As a quick fix, I did `export CC=gcc`

<pre>
[linux1@rhel7testmachine caddy]$ go env

GO111MODULE=""
GOARCH="s390x"
GOBIN=""
GOCACHE="/home/linux1/.cache/go-build"
GOENV="/home/linux1/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="s390x"
GOHOSTOS="linux"
GOINSECURE=""
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/linux1/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_s390x"
GCCGO="/usr/bin/gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/linux1/caddy/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -march=z196 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build949918037=/tmp/go-build -gno-record-gcc-switches"

$ go version
go version go1.14.2 linux/s390x

$  go test   ./...
ok  	github.com/caddyserver/caddy/v2	(cached)
?   	github.com/caddyserver/caddy/v2/caddyconfig	[no test files]
ok  	github.com/caddyserver/caddy/v2/caddyconfig/caddyfile	0.049s
ok  	github.com/caddyserver/caddy/v2/caddyconfig/httpcaddyfile	(cached)
ok  	github.com/caddyserver/caddy/v2/caddytest	(cached)
ok  	github.com/caddyserver/caddy/v2/caddytest/integration	(cached)
?   	github.com/caddyserver/caddy/v2/cmd	[no test files]
?   	github.com/caddyserver/caddy/v2/cmd/caddy	[no test files]
ok  	github.com/caddyserver/caddy/v2/modules/caddyhttp	(cached)
?   	github.com/caddyserver/caddy/v2/modules/caddyhttp/caddyauth	[no test files]
ok  	github.com/caddyserver/caddy/v2/modules/caddyhttp/encode	(cached) [no tests to run]
?   	github.com/caddyserver/caddy/v2/modules/caddyhttp/encode/gzip	[no test files]
?   	github.com/caddyserver/caddy/v2/modules/caddyhttp/encode/zstd	[no test files]
ok  	github.com/caddyserver/caddy/v2/modules/caddyhttp/fileserver	(cached)
ok  	github.com/caddyserver/caddy/v2/modules/caddyhttp/headers	(cached)
?   	github.com/caddyserver/caddy/v2/modules/caddyhttp/requestbody	[no test files]
ok  	github.com/caddyserver/caddy/v2/modules/caddyhttp/reverseproxy	(cached)
ok  	github.com/caddyserver/caddy/v2/modules/caddyhttp/reverseproxy/fastcgi	(cached) [no tests to run]
ok  	github.com/caddyserver/caddy/v2/modules/caddyhttp/rewrite	(cached)
?   	github.com/caddyserver/caddy/v2/modules/caddyhttp/standard	[no test files]
ok  	github.com/caddyserver/caddy/v2/modules/caddyhttp/templates	(cached)
?   	github.com/caddyserver/caddy/v2/modules/caddypki	[no test files]
ok  	github.com/caddyserver/caddy/v2/modules/caddytls	(cached)
?   	github.com/caddyserver/caddy/v2/modules/caddytls/distributedstek	[no test files]
?   	github.com/caddyserver/caddy/v2/modules/caddytls/standardstek	[no test files]
?   	github.com/caddyserver/caddy/v2/modules/filestorage	[no test files]
?   	github.com/caddyserver/caddy/v2/modules/logging	[no test files]
?   	github.com/caddyserver/caddy/v2/modules/standard	[no test files]
</pre>

Note `go test -race` not supported on s390x (ppc64le ok)
<pre>
$ go test -short -race ./...
go test: -race is only supported on linux/amd64, linux/ppc64le, linux/arm64, freebsd/amd64, netbsd/amd64, darwin/amd64 and windows/amd64
</pre>

### Emulation builds of IBM Z / LinuxONE

I haven't been able to yet repeat the error in the build `go: github.com/Masterminds/sprig/v3@v3.0.2: Get "https://proxy.golang.org/github.com/%21masterminds/sprig/v3/@v/v3.0.2.mod": tls: invalid signature by the server certificate: ECDSA verification failure` with the combination of qemu/docker versions, unfortunately I fail at stages before this error.

### Self-hosted runners

Unfortunately self-hosted runners on github workflows require an [agent runner](https://github.com/actions/runner) aren't distributed in ppc64le or s390x forms, and even with the source code, seems to depend on the Microsoft C# SDK for the build.

I haven't totally given up here, a mono build might be possible for ppc64le, there's no s390x build of mono however.

Also looking at binfmt_misc emulation of the x86_64 agent runner on s390x/ppc64le achitecute.

**Comment by Mohammed90** [comment]
Adding to @grooverdan input, I've used the links you provided, I spun an s390x VM and built/tested Caddy with `CGO_ENABLED=0`. It builds and works fine too. I'm exploring and brainstorming a bit. Hopefully we'll come up with something that works.

**Comment by grooverdan** [comment]
The binfmt_misc approach of the running on ppc64le failed quickly as the runner code does CPUid based determination and is seems the C#/netsdk is a JIT. The amount of effort to get a github workflows runner going is far to much for just your project.

Easiest way would be use the real hardware ppc64le and s390x of [travis](https://docs.travis-ci.com/user/multi-cpu-architectures/). I had a little look but couldn't find the definitive motivation for caddy not using them/moving from them so forgive the suggestion if this has been dismissed. It was probably dismissed before the added architecture options where added however. Travis rules are pretty easy however I'm happy to draft something if you want.

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+dist: bionic
+arch:
+  - ppc64le
+  - s390x
+env:
+  - GO111MODULE=on CGO_ENABLED=0
+cache:
+  directories:
+    - $GOPATH/pkg/mod
+    - /home/travis/.cache/go-build
```
[the remark this answers is not in the document]
Comment (Mohammed90) [reply .travis.yml:16]: I don't think there's a way to set the value of this with the output of `go env GOCACHE`. So I snipped this path from reading Travis [logs](https://travis-ci.org/github/caddyserver/caddy/jobs/682411199#L124).

[the remark this answers is not in the document]
Comment (francislavoie) [reply .travis.yml:16]: I guess GOCACHE env isn't set? Either way that looks fine 👍

**Comment by grooverdan** [comment]
Thank you @Mohammed90, @francislavoie  and @mholt for making this vision a reality. Thanks for the support @hairyhenderson.

**Pull request merged**